### ISSUE TITLE
JACK audio on macOS survey

### ISSUE THREAD

**Issue by umlaeute** [pure-data/pure-data#1512]
JACK on macOS is giving me a bit of trouble and i would like to understand what is working and what not.

i'm using this ticket to collect different scenarios and what is working where in a central place (rather than a series of mails on the pd-list that are hard to search afterwards)

See the discussion in the comments

| ID          | System                       | Pd                                                                                                    | JACK                                                                                                                                 | success?                                             |
|-------------|------------------------------|-------------------------------------------------------------------------------------------------------|--------------------------------------------------------------------------------------------------------------------------------------|------------------------------------------------------|
| 20211213jmz | El Capitan (10.11.6) / amd64 | [msp1](http://msp.ucsd.edu/Software/pd-0.52-0test3.macos.zip)                                         | *                                                                                                                                    | no (Pd unusable)                                     |
| 20211213jmz | El Capitan (10.11.6) / amd64 | [iem-ci1](https://git.iem.at/pd/pure-data/-/jobs/33785/artifacts/file/Pd-0.52-0test3-2-g6ca9539f.dmg) | (not installed)                                                                                                                      | yes (sound via portaudio)                            |
| 20211213jmz | El Capitan (10.11.6) / amd64 | [iem-ci1](https://git.iem.at/pd/pure-data/-/jobs/33785/artifacts/file/Pd-0.52-0test3-2-g6ca9539f.dmg) | [jack 0.92_b3](https://github.com/jackaudio/jackaudio.github.com/releases/download/1.9.11/JackOSX.0.92_b3.zip)                       | no (but can be forced to work)    :dizzy_face:                    |
| 20211213jmz | El Capitan (10.11.6) / amd64 | [iem-ci1](https://git.iem.at/pd/pure-data/-/jobs/33785/artifacts/file/Pd-0.52-0test3-2-g6ca9539f.dmg) | [jack2 1.9.19/universal](https://github.com/jackaudio/jack2-releases/releases/download/v1.9.19/jack2-macOS-universal-v1.9.19.tar.gz) | works, but not really usable                         |
| 20211213jmz | El Capitan (10.11.6) / amd64 | [robotcowboy1](http://docs.danomatika.com/pdbuilds/0.52/Pd-0.52-0test3-x86_64-jack.zip)               | (not installed)                                                                                                                      | yes (sound via portaudio)                            |
| 20211213jmz | El Capitan (10.11.6) / amd64 | [robotcowboy1](http://docs.danomatika.com/pdbuilds/0.52/Pd-0.52-0test3-x86_64-jack.zip)               | [jack 0.92_b3](https://github.com/jackaudio/jackaudio.github.com/releases/download/1.9.11/JackOSX.0.92_b3.zip)                       | no (but can be forced to work)    :dizzy_face:                    |
| 20211213jmz | El Capitan (10.11.6) / amd64 | [robotcowboy1](http://docs.danomatika.com/pdbuilds/0.52/Pd-0.52-0test3-x86_64-jack.zip)               | [jack2 1.9.19/universal](https://github.com/jackaudio/jack2-releases/releases/download/v1.9.19/jack2-macOS-universal-v1.9.19.tar.gz) | works, but not really usable                         |
| 20211213wb  | Big Sur (11.6.1) / amd64     | [msp1](http://msp.ucsd.edu/Software/pd-0.52-0test3.macos.zip)                                         | [jack2 1.9.19/amd64](https://github.com/jackaudio/jack2-releases/releases/download/v1.9.19/jack2-macOS-intel-v1.9.19.tar.gz)         | no (running JACK not detected)     :no_entry:                   |
| 20211213wb  | Big Sur (11.6.1) / amd64     | [iem-ci1](https://git.iem.at/pd/pure-data/-/jobs/33785/artifacts/file/Pd-0.52-0test3-2-g6ca9539f.dmg) | [jack2 1.9.19/amd64](https://github.com/jackaudio/jack2-releases/releases/download/v1.9.19/jack2-macOS-intel-v1.9.19.tar.gz)         | no (running JACK not detected)     :no_entry:                   |
| 20211213wb  | Big Sur (11.6.1) / amd64     | [robotcowboy1](http://docs.danomatika.com/pdbuilds/0.52/Pd-0.52-0test3-x86_64-jack.zip)               | [jack2 1.9.19/amd64](https://github.com/jackaudio/jack2-releases/releases/download/v1.9.19/jack2-macOS-intel-v1.9.19.tar.gz)         | yes :tada:                                           |
| 20211216rdz | Catalina (10.15.7) / amd64   | [msp1](http://msp.ucsd.edu/Software/pd-0.52-0test3.macos.zip)                                         | [jack 0.92_b3](https://github.com/jackaudio/jackaudio.github.com/releases/download/1.9.11/JackOSX.0.92_b3.zip)                       | `Can't open Jack (it seems not to be installed)`   :dizzy_face:   |
| 20211216rdz | Catalina (10.15.7) / amd64   | [msp 0.52-test4](http://msp.ucsd.edu/Software/pd-0.52-0test4.macos.zip)                               | [jack 0.92_b3](https://github.com/jackaudio/jackaudio.github.com/releases/download/1.9.11/JackOSX.0.92_b3.zip)                       | `Can't open Jack (it seems not to be installed)`  :dizzy_face:    |
| 20211216rdz | Catalina (10.15.7) / amd64   | [msp 0.51-4 ](http://msp.ucsd.edu/Software/pd-0.51-4.mac.tar.gz)                                      | [jack 0.92_b3](https://github.com/jackaudio/jackaudio.github.com/releases/download/1.9.11/JackOSX.0.92_b3.zip)                       | works                                                |
| 20211216rdz | Catalina (10.15.7) / amd64   | [iem-ci1](https://git.iem.at/pd/pure-data/-/jobs/33785/artifacts/file/Pd-0.52-0test3-2-g6ca9539f.dmg) | [jack 0.92_b3](https://github.com/jackaudio/jackaudio.github.com/releases/download/1.9.11/JackOSX.0.92_b3.zip)                       | `Can't open Jack (it seems not to be installed)`    :dizzy_face:  |
| 20211216rdz | Catalina (10.15.7) / amd64   | [iem-ci2](https://git.iem.at/pd/pure-data/-/jobs/33818/artifacts/file/Pd-0.52-0test3-2-g6ca9539f.dmg) | [jack 0.92_b3](https://github.com/jackaudio/jackaudio.github.com/releases/download/1.9.11/JackOSX.0.92_b3.zip)                       | `Can't open Jack (it seems not to be installed)`    :dizzy_face:  |
| 20211216rdz | Catalina (10.15.7) / amd64   | [robotcowboy1](http://docs.danomatika.com/pdbuilds/0.52/Pd-0.52-0test3-x86_64-jack.zip)               | [jack 0.92_b3](https://github.com/jackaudio/jackaudio.github.com/releases/download/1.9.11/JackOSX.0.92_b3.zip)                       | `Can't open Jack (it seems not to be installed)`    :dizzy_face:  |
| 20211216rdz | Catalina (10.15.7) / amd64   | [msp 0.51-4](http://msp.ucsd.edu/Software/pd-0.51-4.mac.tar.gz)                                       | [jack2 1.9.19/amd64](https://github.com/jackaudio/jack2-releases/releases/download/v1.9.19/jack2-macOS-intel-v1.9.19.tar.gz)         | `JACK: couldn't connect to server, is JACK running?` :no_entry:  |
| 20211216rdz | Catalina (10.15.7) / amd64   | [iem-ci1](https://git.iem.at/pd/pure-data/-/jobs/33785/artifacts/file/Pd-0.52-0test3-2-g6ca9539f.dmg) | [jack2 1.9.19/amd64](https://github.com/jackaudio/jack2-releases/releases/download/v1.9.19/jack2-macOS-intel-v1.9.19.tar.gz)         | `Can't open Jack (it seems not to be installed)`    :dizzy_face:  |
| 20211216rdz | Catalina (10.15.7) / amd64   | [iem-ci2](https://git.iem.at/pd/pure-data/-/jobs/33818/artifacts/file/Pd-0.52-0test3-2-g6ca9539f.dmg) | [jack2 1.9.19/amd64](https://github.com/jackaudio/jack2-releases/releases/download/v1.9.19/jack2-macOS-intel-v1.9.19.tar.gz)         | works                                                |
| 20211216rdz | Catalina (10.15.7) / amd64   | [msp1](http://msp.ucsd.edu/Software/pd-0.52-0test3.macos.zip)                                         | [jack2 1.9.19/amd64](https://github.com/jackaudio/jack2-releases/releases/download/v1.9.19/jack2-macOS-intel-v1.9.19.tar.gz)         | works                                                |
| 20211216rdz | Catalina (10.15.7) / amd64   | [msp 0.52-test4](http://msp.ucsd.edu/Software/pd-0.52-0test4.macos.zip)                               | [jack2 1.9.19/amd64](https://github.com/jackaudio/jack2-releases/releases/download/v1.9.19/jack2-macOS-intel-v1.9.19.tar.gz)         | works                                                |
| 20211216rdz | Catalina (10.15.7) / amd64   | [robotcowboy1](http://docs.danomatika.com/pdbuilds/0.52/Pd-0.52-0test3-x86_64-jack.zip)               | [jack2 1.9.19/amd64](https://github.com/jackaudio/jack2-releases/releases/download/v1.9.19/jack2-macOS-intel-v1.9.19.tar.gz)         | works                                                |

#### Pure Data builds

- Pd "msp1"
  - Miller's Pd-0.52-0test3 ["compiled for Macintosh OS 10 or later"](http://msp.ucsd.edu/Software/pd-0.52-0test3.macos.zip)
- Pd "iem-ci1"
  - 6ca9539fe0100dc0a0b4203b637f91e0eed2a84d
  - downloaded from [iem-ci](https://git.iem.at/pd/pure-data/-/jobs/33785/artifacts/file/Pd-0.52-0test3-2-g6ca9539f.dmg)
  - built on IEM-CI on 2021-12-13 10:37
    - *macOS Mojave (10.14)*
    - [logs](https://git.iem.at/pd/pure-data/-/jobs/33771)
    - dependencies
      - `brew install jack` (installed https://ghcr.io/v2/homebrew/core/jack/manifests/1.9.19-2)
      -  Tcl/Tk-8.6.12 universal test-build by miller (http://msp.ucsd.edu/tmp/Wish-8.6.12.app.tgz)
    - configured with `--disable-jack-framework --enable-jack`
 
- Pd "iem-ci2"
  - 6ca9539fe0100dc0a0b4203b637f91e0eed2a84d
  - downloaded from [iem-ci](https://git.iem.at/pd/pure-data/-/jobs/33818/artifacts/file/Pd-0.52-0test3-2-g6ca9539f.dmg)
  - built on IEM-CI on 2021-12-13 22:01
    - *macOS Mojave (10.14)*
    - [logs](https://git.iem.at/pd/pure-data/-/jobs/33818)
    - dependencies
      - JACK installed from [jackaudio.org](https://github.com/jackaudio/jack2-releases/releases/download/v1.9.19/jack2-macOS-universal-v1.9.19.tar.gz)
      -  Tcl/Tk-8.6.12 universal test-build by miller (http://msp.ucsd.edu/tmp/Wish-8.6.12.app.tgz)
    - configured with `--disable-jack-framework --enable-jack`

- Pd "robotcowboy1"
  - downloaded from http://docs.danomatika.com/pdbuilds/0.52/Pd-0.52-0test3-x86_64-jack.zip
    - dependencies
      - built with "JACK from https://jackaudio.org"
    - configured with `--disable-jack-framework --enable-jack`

**Comment by umlaeute** [comment]
### OS X El Capitan (10.11.6)

- Pd
  - 6ca9539fe0100dc0a0b4203b637f91e0eed2a84d
  - downloaded from [iem-ci ](https://git.iem.at/pd/pure-data/-/jobs/33785/artifacts/file/Pd-0.52-0test3-2-g6ca9539f.dmg)
  - built on IEM-CI on 2021-12-13 10:37
    - *macOS Mojave (10.14)*
    - [logs](https://git.iem.at/pd/pure-data/-/jobs/33771)
    - dependencies
      - `brew install jack` (installed https://ghcr.io/v2/homebrew/core/jack/manifests/1.9.19-2)
      -  Tcl/Tk-8.6.12 test-build by miller (http://msp.ucsd.edu/tmp/Wish-8.6.12.app.tgz)
    - built with `--disable-jack-framework --enable-jack`
- JACK
  - release 1.9.19
  - downloaded from [https://jackaudio.org/downloads/](https://github.com/jackaudio/jack2-releases/releases/download/v1.9.19/jack2-macOS-universal-v1.9.19.tar.gz
- JackOSX
  - release 0.92_b3 (2017)
  - downloaded from [https://web.archive.org/web/20200512230754/https://jackaudio.org/downloads/](https://github.com/jackaudio/jackaudio.github.com/releases/download/1.9.11/JackOSX.0.92_b3.zip)

#### results

| Pd | JACK | success? |
|-----|---------|--------------|
|6ca9539fe0100dc0a0b4203b637f91e0eed2a84d| (not installed) | yes (sound via portaudio) |
|6ca9539fe0100dc0a0b4203b637f91e0eed2a84d|1.9.19| no[¹](#) |
|6ca9539fe0100dc0a0b4203b637f91e0eed2a84d|0.92_b3| yes[²](#) |

##### ¹ details JACK-1.9.19

According to the [homepage](https://jackaudio.org/downloads/) the pre-built binaries are for "macOS 10.12 or higher."
And indeed the embedded ´QJackCtl` cannot be started.

Starting jackd manually (`jackd -d coreaudio`) stalls Pd, and the `jackd` process gives errors like:

```sh
$ jackd -ddummy    
jackdmp 1.9.19
Copyright 2001-2005 Paul Davis and others.
Copyright 2004-2016 Grame.
Copyright 2016-2021 Filipe Coelho.
jackdmp comes with ABSOLUTELY NO WARRANTY
This is free software, and you are welcome to redistribute it
under certain conditions; see the file COPYING for details
JACK server starting in realtime mode with priority 10
self-connect-mode is "Don't restrict self connect requests"
BDB1546 unable to join the environment
Cannot open DB environment: Resource temporarily unavailable
BDB1546 unable to join the environment
Cannot open DB environment: Resource temporarily unavailable
BDB1546 unable to join the environment
Cannot open DB environment: Resource temporarily unavailable
BDB1546 unable to join the environment
Cannot open DB environment: Resource temporarily unavailable
BDB1546 unable to join the environment
Cannot open DB environment: Resource temporarily unavailable
^CJack main caught signal 2
```

##### ² details JackOSX-0.92_b3

This rather old release provides a `/usr/local/lib/libjack.0.dylib` (and /usr/local/lib/libjack.dylib`).
However, Pd links against `/usr/local/lib/libjack.0.1.0.dylib` (which this package does not provide).
Using a hacky symlink, allows us to use JACK just fine (if we enabled `callbacks`):

~~~sh
$ ls -nlha /usr/local/lib/libjack.*dylib
lrwxr-xr-x  1 0    80    15B Dec 13 14:29 /usr/local/lib/libjack.0.1.0.dylib -> libjack.0.dylib
lrwxr-xr-x  1 501  80    43B Dec 13 14:23 /usr/local/lib/libjack.0.dylib -> /Library/Frameworks/Jackmp.framework/Jackmp
lrwxr-xr-x  1 501  80    43B Dec 13 14:23 /usr/local/lib/libjack.dylib -> /Library/Frameworks/Jackmp.framework/Jackmp
~~~

**Comment by wbrent** [comment]
# macOS Big Sur (11.6.1)

### System
- 2017 MacBook Pro (Intel 64bit) running macOS Big Sur (11.6.1)

### JACK
- Version 1.9.19 installed via the Intel 64bit Installer from https://jackaudio.org/downloads/ 

### Pure Data

- Pd "msp"
    - Downloaded from <a href="http://msp.ucsd.edu/software.html">Miller's site</a> "compiled for Macintosh OS 10 or later"
    - This fails to connect to the running JACK server
    ```
    $ ./Pd-0.52-0test3.app/Contents/Resources/bin/pd -jack -callback Pd-0.52-0test3.app/Contents/Resources/doc/7.stuff/tools/testtone.pd
    JACK protocol mismatch 8
    Cannot connect to the server
    JackShmReadWritePtr1::~JackShmReadWritePtr1 - Init not done for -1, skipping unlock
    JackShmReadWritePtr::~JackShmReadWritePtr - Init not done for -1, skipping unlock
    JackShmReadWritePtr::~JackShmReadWritePtr - Init not done for -1, skipping unlock
    ```
    - The Pd console shows error "JACK: couldn't connect to server, is JACK running?"

- Pd "iem-ci"
    - Downloaded from <a href="https://git.iem.at/pd/pure-data/-/jobs/33785/artifacts/file/Pd-0.52-0test3-2-g6ca9539f.dmg">iem-ci</a>
    - This fails to connect to the running JACK server
    ```
    $ ./Pd-0.52-0test3-2-g6ca9539f.app/Contents/Resources/bin/pd -jack -callback Pd-0.52-0test3-2-g6ca9539f.app/Contents/Resources/doc/7.stuff/tools/testtone.pd
    JACK protocol mismatch 8
    Cannot connect to the server
    JackShmReadWritePtr1::~JackShmReadWritePtr1 - Init not done for -1, skipping unlock
    JackShmReadWritePtr::~JackShmReadWritePtr - Init not done for -1, skipping unlock
    JackShmReadWritePtr::~JackShmReadWritePtr - Init not done for -1, skipping unlock
    ```
    - The Pd console shows error "JACK: couldn't connect to server, is JACK running?"

- Pd "robotcowboy"
    - Downloaded from http://docs.danomatika.com/pdbuilds/0.52/Pd-0.52-0test3-x86_64-jack.zip
    - This successfully connects to the running JACK server

**Comment by reduzent** [comment]
# macOS Catalina (10.15.7)

### System
  * Hardware: **MacBook Pro (2016)**
  * OS: macOS Catalina (10.15.7)


| Pd build | JACK build | Result |
|-----------|-----------|---------------|
| [msp1](http://msp.ucsd.edu/Software/pd-0.52-0test3.macos.zip) |  jack 0.92_b3    |  `Can't open Jack (it seems not to be installed)` |
| [msp 0.52-test4](http://msp.ucsd.edu/Software/pd-0.52-0test4.macos.zip) |  jack 0.92_b3  |  `Can't open Jack (it seems not to be installed)` |
| [msp 0.51-4 ](http://msp.ucsd.edu/Software/pd-0.51-4.mac.tar.gz) |     jack 0.92_b3  |     works |
| [iemci](https://git.iem.at/pd/pure-data/-/jobs/33785/artifacts/file/Pd-0.52-0test3-2-g6ca9539f.dmg)   |         jack 0.92_b3   |    `Can't open Jack (it seems not to be installed)` |
| [robotcowboy1](http://docs.danomatika.com/pdbuilds/0.52/Pd-0.52-0test3-x86_64-jack.zip)  |   jack 0.92_b3  |     `Can't open Jack (it seems not to be installed)` |
| [msp 0.51-4](http://msp.ucsd.edu/Software/pd-0.51-4.mac.tar.gz)   |   [jack2 1.9.19/amd64](https://github.com/jackaudio/jack2-releases/releases/download/v1.9.19/jack2-macOS-intel-v1.9.19.tar.gz)  |  `JACK: couldn't connect to server, is JACK running?` |
| [iemci](https://git.iem.at/pd/pure-data/-/jobs/33785/artifacts/file/Pd-0.52-0test3-2-g6ca9539f.dmg)   |      [jack2 1.9.19/amd64](https://github.com/jackaudio/jack2-releases/releases/download/v1.9.19/jack2-macOS-intel-v1.9.19.tar.gz) |  `Can't open Jack (it seems not to be installed)` |
| [msp1](http://msp.ucsd.edu/Software/pd-0.52-0test3.macos.zip) |         [jack2 1.9.19/amd64](https://github.com/jackaudio/jack2-releases/releases/download/v1.9.19/jack2-macOS-intel-v1.9.19.tar.gz)  |  works[^1] |
| [msp 0.52-test4](http://msp.ucsd.edu/Software/pd-0.52-0test4.macos.zip) |  [jack2 1.9.19/amd64](https://github.com/jackaudio/jack2-releases/releases/download/v1.9.19/jack2-macOS-intel-v1.9.19.tar.gz) |  works[^1] |
| [robotcowboy1](http://docs.danomatika.com/pdbuilds/0.52/Pd-0.52-0test3-x86_64-jack.zip)   |   [jack2 1.9.19/amd64](https://github.com/jackaudio/jack2-releases/releases/download/v1.9.19/jack2-macOS-intel-v1.9.19.tar.gz) |   works[^1] |


**NOTE**: None of the tested Pd builds works with both, JACK1 (0.92_b3) and JACK2 (1.9.19)

[^1]: works reliably with 'use callbacks' unchecked and checked.

**Comment by umlaeute** [comment]
it should be possible to fix the *not working* on "jack 0.92_b3" as described in my [`details JackOSX-0.92_b3`](https://github.com/pure-data/pure-data/issues/1512#issuecomment-992524482). that's what I meant with "yes (with quirks)". i'll probably re-phrase this to "no (but can be convinced to work)"

**Comment by umlaeute** [comment]
afaik, the JACK issues have been resolved for macOS. so I'm closing this.
